src/libguac/guacamole/client.h

```diff
@@ -162,14 +173,65 @@ struct guac_client {
      * Lock which is acquired when the users list is being manipulated, or when
      * the users list is being iterated.
      */
-    pthread_rwlock_t __users_lock;
+    guac_local_lock __users_lock;
 
     /**
      * The first user within the list of all connected users, or NULL if no
      * users are currently connected.
      */
     guac_user* __users;
 
+    /**
+     * Lock which is acquired when the pending users list is being manipulated,
+     * or when the pending users list is being iterated.
+     */
+    guac_local_lock __pending_users_lock;
+
+    /**
+     * A timer that will periodically synchronize the list of pending users,
+     * emptying the list once synchronization is complete. Only for internal
+     * use within the client. This will be NULL until the first user joins
+     * the connection, as it is lazily instantiated at that time.
+     */
+    timer_t __pending_users_timer;
+
+    /**
+     * Non-zero if the pending users timer is configured and running, or zero
+     * otherwise.
+     */
+    int __pending_users_timer_running;
+
+    /**
+     * A mutex that must be acquired before modifying the pending users timer.
+     */
+    pthread_mutex_t __pending_users_timer_mutex;
+
+    /**
+     * Notification / signaling configuration for the pending users timer.
+     * Only for internal use within the client. This will be NULL until the
+     * first user joins the connection.
+     */
+    struct sigevent __pending_users_timer_signal_config;
+
+    /**
+     * Specification for the timing of the pending users timer. Only for
+     * internal use within the client. This will be NULL until the
+     * first user joins the connection.
+     */
+    struct itimerspec __pending_users_time_spec;
```

Comment:
Do these need to be included within the `guac_client` structure? I see them used only during initialization.

Reply:
Probably not? When constructing the timer, you pass pointers to these various configuration-time things. The API docs don't say or even imply whether you need to keep them around afterwards, though it seems unlikely that any implementation would actually expect them to stick around for the whole timer lifetime.

I can just allocate them on the stack and not worry about it if it works correctly in my dev environment, I guess.